### DISCUSSION TITLE
Remove variable types and implement StringSubstitutor

### DISCUSSION THREAD

**Discussion by josefd8** [veepee-oss/gingerspec#61] [Ideas]
Hey @oleksandr0001 

Currently, there are 4 types of variables supported by GingerSpec

- ${} -> Global variables
- !{} -> Local variables
- #{} -> values from property file
- @{} -> special functions

I believe that this could bring confusion to QAs, because sometimes is not clear when to use one or the other. I believe this could all be simplified by using [StringSubstitutor](https://commons.apache.org/proper/commons-text/apidocs/org/apache/commons/text/StringSubstitutor.html) and make ALL variables use the ${} placeholder

- ${} and !{} would be pretty straightforward, I just have to indicate StringSubstitutor to use all System and Thread variables for the string substitution
 
- For #{}, StringSubstitutor has already a function to read variables from properties files, like this 
 `${properties:src/test/resources/document.properties::mykey}`

- For @{}, StringSubstitutor can also perform similar operations, and many more:

```
1. Base64 Encoder:        ${base64Encoder:HelloWorld!}
2. Java Constant:         ${const:java.awt.event.KeyEvent.VK_ESCAPE}
3. Date:                  ${date:yyyy-MM-dd}
4. DNS:                   ${dns:address|apache.org}
5. Environment Variable:  ${env:USERNAME}
6. File Content:          ${file:UTF-8:src/test/resources/document.properties}
7. Java:                  ${java:version} 
8. Localhost:             ${localhost:canonical-name}
9. Properties File:       ${properties:src/test/resources/document.properties::mykey}
10. Resource Bundle:       ${resourceBundle:org.example.testResourceBundleLookup:mykey}
11. Script:                ${script:javascript:3 + 4} 
12. System Property:       ${sys:user.dir}
13. URL Decoder:           ${urlDecoder:Hello%20World%21}
14. URL Encoder:           ${urlEncoder:Hello World!}
15. URL Content (HTTP):    ${url:UTF-8:http://www.apache.org}
16. URL Content (HTTPS):   ${url:UTF-8:https://www.apache.org}
17. URL Content (File):    ${url:UTF-8:file:///${sys:user.dir}/src/test/resources/document.properties}
18. XML XPath:             ${xml:src/test/resources/document.xml:/root/path/to/node}
```

I'm thinking about implementing this code but leaving the rest of the operators still working and show a warning when the user runs the test and a !{}, @{} or #{} variable is used, asking them to comply with the new format.

**Comment by josefd8** [comment]
By the way, this library also supports using default values 

`${name:-jose}`

And also has support for nested variables:

`${script:javascript:'${name}'.toUpperCase()}`

**Reply by oleksandr0001** [reply]
@josefd8 it sounds very good. Does anyone keep using "special functions"? Probably is not necessary to spend time on this (@{}). 
Does all this suppose any change for the QA project? (location of properties, execution parameters)

**Reply by josefd8** [reply]
@oleksandr0001 I really cannot remember at this moment anyone using @{}, but I believe part of the reason is because the documentation for this is not very clear and because it uses a different placeholder than other variables (${} !{})

I'm not expecting this to cause changes to QA projects since I'm planning to leave the old code still in place, so 'old' variables would still work.

One cool thing of StringSubstitutor is that you can create your own string lookups, for example, I created the following ones:

${toUpperCase:myVar} -> transform variable to upper case (outputs MYVAR)
${toLowerCase:myVar} -> transform variable to lower case (outputs myvar)
${envProperties:myVar} -> gets the myVar property from the properties files located under etc/test/resources. This is the equivalent as the old #{myVar}
